rfc/004-reverse-sync.md

```diff
@@ -0,0 +1,198 @@
+# RFC 004: ReverseSync - fetching historical data
+
+## Changelog
+
+- 2021-03-03: Simplify proposal to the state sync case.
+- 2021-02-17: Add notes on asynchronicity of processes.
+- 2020-12-10: Rename backfill blocks to reverse sync.
+- 2020-11-25: Initial draft.
+
+## Author(s)
+
+- Callum Waters (@cmwaters)
+
+## Context
+
+Two new features: [Block pruning](https://github.com/tendermint/tendermint/issues/3652)
+and [State sync](https://github.com/tendermint/tendermint/blob/master/docs/architecture/adr-042-state-sync.md)
+meant nodes no longer needed a complete history of the blockchain. This
+introduced some challenges of its own which were covered and subsequently
+tackled with [RFC-001](https://github.com/tendermint/spec/blob/master/rfc/001-block-retention.md).
+The RFC allowed applications to set a block retention height; an upper bound on
+what blocks would be pruned. However nodes who state sync past this upper bound
+(which is necessary as snapshots must be saved within the trusting period for
+the assisting light client to verify) have no means of backfilling the blocks
+to meet the retention limit. This could be a problem as nodes who state sync and
+then eventually switch to consensus may not have the block and validator
+history to verify evidence causing them to panic if they see 2/3 commit on what
+the node believes to be an invalid block.
+
+Thus, this RFC sets out to instil a minimum block history invariant amongst
+honest nodes.
+
+## Proposal
+
+A backfill mechanism can simply be defined as an algorithm for fetching,
+verifying and storing, headers and validator sets of a height prior to the
+current base of the node's blockchain. In matching the terminology used for
+other data retrieving protocols (i.e. fast sync and state sync), we
+call this method **ReverseSync**.
+
+We will define the mechanism in four sections:
+- Usage
+- Verification
+- Design
+- Termination
+
+### Usage
+
+For now, we focus purely on the case of a state syncing node, whom after
+syncing to a height will need to verify an adequate amount of previous data in
+order to be capable of processing new blocks and advancing sate. We can call
```

Comment:
I find this section a bit confusing and hard to conceptualize. Perhaps it's just me. 

> ... will need to verify an adequate amount of previous data in
order to be capable of processing new blocks and advancing sate.

What is adequate? What data exactly? I would state that we need data for blocks from the state sync snapshot back until the max evidence age, in order to process evidence and auxiliary information.

Reply:
In any case, I think I can try reword this better